src/D3MHub.sol

```diff
@@ -0,0 +1,601 @@
+    function skim(bytes32, address) external;
+}
+
+interface DaiJoinLike {
+    function dai() external view returns (address);
+    function join(address, uint256) external;
+    function exit(address, uint256) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external returns (bool);
+}
+
+/**
+    @title D3M Hub
+    @notice This is the main D3M contract and is responsible for winding and
+    unwinding pools, interacting with DSS and tracking the plans and pools and
+    their states.
+*/
+contract D3MHub {
+
+    // --- Auth ---
+    /**
+        @notice Maps address that have permission in the Pool.
+        @dev 1 = allowed, 0 = no permission
+        @return authorization 1 or 0
+    */
+    mapping (address => uint256) public wards;
+
+    address public vow;
+    EndLike public end;
+    uint256 public locked;
+
+    /// @notice maps ilk bytes32 to the D3M tracking struct.
+    mapping (bytes32 => Ilk) public ilks;
+
+    VatLike     public immutable vat;
+    DaiJoinLike public immutable daiJoin;
+
+    enum Mode { NORMAL, MODULE_CULLED, MCD_CAGED }
+
+    /**
+        @notice Tracking struct for each of the D3M ilks.
+        @param pool   Contract to access external pool and hold balances
+        @param plan   Contract used to calculate target debt
+        @param tau    Time until you can write off the debt [sec]
+        @param culled Debt write off triggered (1 or 0)
+        @param tic    Timestamp when the pool is caged
+    */
+    struct Ilk {
+        ID3MPool pool;   // Access external pool and holds balances
+        ID3MPlan plan;   // How we calculate target debt
+        uint256  tau;    // Time until you can write off the debt [sec]
+        uint256  culled; // Debt write off triggered
+        uint256  tic;    // Timestamp when the d3m can be culled (tau + timestamp when caged)
+    }
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, address data);
+    event File(bytes32 indexed ilk, bytes32 indexed what, address data);
+    event File(bytes32 indexed ilk, bytes32 indexed what, uint256 data);
+    event Wind(bytes32 indexed ilk, uint256 amount);
+    event Unwind(bytes32 indexed ilk, uint256 amount, uint256 fees);
+    event Reap(bytes32 indexed ilk, uint256 amt);
+    event Exit(bytes32 indexed ilk, address indexed usr, uint256 amt);
+    event Cage(bytes32 indexed ilk);
+    event Cull(bytes32 indexed ilk, uint256 ink, uint256 art);
+    event Uncull(bytes32 indexed ilk, uint256 wad);
+
+    /**
+        @dev sets msg.sender as authed.
+        @param vat_     address of the DSS vat contract
+        @param daiJoin_ address of the DSS Dai Join contract
+    */
+    constructor(address vat_, address daiJoin_) {
+        vat = VatLike(vat_);
+        daiJoin = DaiJoinLike(daiJoin_);
+        TokenLike(DaiJoinLike(daiJoin_).dai()).approve(daiJoin_, type(uint256).max);
```

Comment:
No need to repeatidly cast these after Solidity 0.8. Can just do `TokenLike(daiJoin.dai()).approve(daiJoin_, type(uint256).max);`

Reply:
0b42cfa